lib/engine/game/base.rb

```diff
@@ -1924,30 +1924,62 @@ def ability_right_owner?(entity, ability, owner_type)
           when nil
             true
           end
-        return false unless correct_owner_type
-        return false if owner_type && (ability.owner_type.to_s != owner_type.to_s)
 
-        true
+        !!correct_owner_type
       end
 
       def ability_usable_this_or?(ability)
         !ability.count_per_or || (ability.count_this_or < ability.count_per_or)
       end
 
-      def ability_right_time?(ability, time, strict_time)
-        return false if strict_time && !ability.when
-        return true unless time
+      def ability_right_time?(ability, time, allow_passive)
+        if (times = Array(time).map(&:to_s)).size > 1
+          return times.any? { |t| ability_right_time?(ability, t, allow_passive) }
+        end
+
+        return true if time == 'all'
+        return true unless @round
 
-        if (ability.type == :tile_lay) && (step = ability_blocking_step)&.is_a?(Step::SpecialTrack)
-          return step.company == ability.owner
+        current_step = ability_blocking_step
+        current_step_name = current_step.class.name.split('::').last
+        if (ability.type == :tile_lay) && current_step&.is_a?(Step::SpecialTrack)
+          return current_step.company == ability.owner
         end
 
-        if ability.when == 'any'
-          !strict_time
-        elsif ability.when == 'owning_corp_or_turn'
-          %w[owning_corp_or_turn track].include?(time)
-        else
-          ability.when == time.to_s
+        time = current_step_name if current_step && (time == '%current_step%')
+
+        allow_passive = true if !time && allow_passive.nil?
+        return false if ability.passive && !allow_passive
+        return true if ability.passive && allow_passive && ability.when.empty?
+
+        return true if ability.when?('any')
+
+        return false if @round.operating? &&
+                        ability.owner.owned_by_corporation? &&
+                        (@round.current_operator != ability.corporation) &&
+                        !ability.when?('other_or')
+
+        return false if time && !ability.when?(time)
+
+        ability.when.any? do |ability_when|
+          case ability_when
```

Comment:
This `case` block was kind of the main idea I had for this PR, with an ability listing out strings that describe when it can be used, and then bits of code to tell when that string matches the current game state.